### PR TITLE
ObjcRuntime.h: Add mips64, aarch64, and riscv64 to non-legacy dispatch

### PR DESCRIPTION
**Pull request by hmelder**
This PR updates the list of architectures for which libobjc2 has fast-path objc_msgSend implementations.

Related to: https://github.com/gnustep/libobjc2/pull/261

### CODE REVIEW

**Comment by hmelder** [comment]
CC @davidchisnall

File: clang/include/clang/Basic/ObjCRuntime.h [hunk 1]
```diff
@@ -101,10 +101,17 @@ class ObjCRuntime {
     // The GNUstep runtime uses a newer dispatch method by default from
     // version 1.6 onwards
     if (getKind() == GNUstep && getVersion() >= VersionTuple(1, 6)) {
-      if (Arch == llvm::Triple::arm ||
-          Arch == llvm::Triple::x86 ||
-          Arch == llvm::Triple::x86_64)
+      switch (Arch) {
```
Comment (davidchisnall) [inline clang/include/clang/Basic/ObjCRuntime.h:104]: The version check above is not quite right for these.

AArch64 support is in 1.9.
RV64 in 2.2.

I think mips64 is correct (it looks like I added the code 9 years ago, which probably lines up with the 1.6 release).

**Comment by davidchisnall** [comment]
This should also come with a test to make sure that we’re generating objc_msgSend or the lookup functions with the runtime versions after / before the support for these architectures, respectively.

File: clang/include/clang/Basic/ObjCRuntime.h (hunk 1 above)
Comment (hmelder) [reply clang/include/clang/Basic/ObjCRuntime.h:104]: > The version check above is not quite right for these.

This was my first approach, but it seems like GNUstep and libobjc2 unit tests are always emitting -fobjc-runtime=gnustep-2.0

So one would need to change this flag in the build configuration everytime libobjc2 gets updated.

Comment (hmelder) [reply clang/include/clang/Basic/ObjCRuntime.h:104]: I can do the version check properly and update the libobjc2 README to properly document this. I probably need to rewrite the detection in gnustep-make too -_-

**Comment by hmelder** [comment]
> This should also come with a test to make sure that we’re generating objc_msgSend or the lookup functions with the runtime versions after / before the support for these architectures, respectively.

Is libobjc2 included in the LLVM CI? I cannot find the place in the workflow where it is installed.

File: clang/include/clang/Basic/ObjCRuntime.h (hunk 1 above)
Comment (davidchisnall) [reply clang/include/clang/Basic/ObjCRuntime.h:104]: > This was my first approach, but it seems like GNUstep and libobjc2 unit tests are always emitting -fobjc-runtime=gnustep-2.0

That's fine.  We should bump that in both places to the correct version.  It's better than having the compiler emit things that are not supported when targeting a specific version of the runtime.

**Comment by davidchisnall** [comment]
No, clang tests are not executable.  This test is the right approximate shape:


https://github.com/llvm/llvm-project/blob/a3e8e86fb6ad27fe070bb2c0f54a1c697c665f13/clang/test/CodeGenObjC/messages.m#L18

The RUN line invokes clang in a few different configurations (you want to invoke it with the `-fobjc-runtime=gnustep-{version}` flag and the correct triples and then check that we get either `objc_msgSend` or `objc_msg_lookup`, depending on whether `objc_msgSend` is expected to be supported in that configuration.

**Comment by hmelder** [comment]
It seems like isLegacyDispatchDefaultForArch is only called by the driver, and not when passing `-cc1`.

https://github.com/llvm/llvm-project/blob/11ac97c67a9315dce48fd938d68ae991e3559f10/clang/lib/Driver/ToolChains/Clang.cpp#L3935-L3953

Using the driver instead of the frontend does not work as there are platform checks in `Clang::AddObjCRuntimeArgs`.

Example on macOS:
```sh
clang: error: GNUstep Objective-C runtime version 2 incompatible with target binary format
```

The result is that without specifying `-fobjc-dispatch-method=mixed` or `-fobjc-dispatch-method=no-legacy`, the frontend will always emit `call ptr @objc_msg_lookup_sender`.

Do you have an idea on how we can test this instead?

Tested with

```sh
#flags="-fobjc-dispatch-method=mixed"
flags=""
rt=gnustep-2.0
triple="x86_64-unknown-freebsd"

build-llvm/bin/clang -cc1\
		-triple ${triple} \
		-fobjc-runtime=${rt} ${flags} \
		-emit-llvm -o - gnustep2-message-dispatch.m | less
```

<details>
<summary>gnustep2-message-dispatch.m</summary>

```objc
// DEFINE: %{triple} =
// DEFINE: %{ver} =
// DEFINE: %{prefix} = CHECK-MSG-SEND
// DEFINE: %{check} = %clang_cc1 -triple %{triple} -fobjc-runtime=gnustep-%{ver} -emit-llvm -o - %s | FileCheck %s -check-prefix %{prefix}


// REDEFINE: %{ver} = 1.6
// REDEFINE: %{triple} = x86_64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = arm-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{prefix} = CHECK-MSG-LOOKUP
// REDEFINE: %{triple} = aarch64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = mips64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = riscv64-unknown-freebsd
// RUN: %{check}

// REDEFINE: %{ver} = 1.9
// REDEFINE: %{prefix} = CHECK-MSG-SEND
// REDEFINE: %{triple} = x86_64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = arm-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = aarch64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = mips64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{prefix} = CHECK-MSG-LOOKUP
// REDEFINE: %{triple} = riscv64-unknown-freebsd
// RUN: %{check}

// REDEFINE: %{ver} = 2.2
// REDEFINE: %{prefix} = CHECK-MSG-SEND
// REDEFINE: %{triple} = x86_64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = arm-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = aarch64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = mips64-unknown-freebsd
// RUN: %{check}
// REDEFINE: %{triple} = riscv64-unknown-freebsd
// RUN: %{check}

typedef struct {
  int x;
  int y;
  int z[10];
} MyPoint;

void f0(id a) {
  int i;
  MyPoint pt = { 1, 2};

  // CHECK-MSG-SEND: call {{.*}} @objc_msgSend
  // CHECK-MSG-LOOKUP: call {{.*}} @objc_msg_lookup(
  [a print0];

  // CHECK-MSG-SEND: call {{.*}} @objc_msgSend
  // CHECK-MSG-LOOKUP: call {{.*}} @objc_msg_lookup(
  [a print1: 10];

  // CHECK-MSG-SEND: call {{.*}} @objc_msgSend
  // CHECK-MSG-LOOKUP: call {{.*}} @objc_msg_lookup(
  [a print2: 10 and: "hello" and: 2.2];

  // CHECK-MSG-SEND: call {{.*}} @objc_msgSend
  // CHECK-MSG-LOOKUP: call {{.*}} @objc_msg_lookup(
  [a takeStruct: pt ];
}
```

</details>

**Comment by hmelder** [comment]
I assume we can split the test into two parts:
1. Checking if the driver correctly sets `-fobjc-dispatch-method`
2. Checking if frontend emits `objc_msgSend`

**Comment by hmelder** [comment]
Definition of -fobjc-dispatch-method:

```
def fobjc_dispatch_method_EQ : Joined<["-"], "fobjc-dispatch-method=">,
  HelpText<"Objective-C dispatch method to use">,
  Values<"legacy,non-legacy,mixed">,
  NormalizedValuesScope<"CodeGenOptions">, NormalizedValues<["Legacy", "NonLegacy", "Mixed"]>,
  MarshallingInfoEnum<CodeGenOpts<"ObjCDispatchMethod">, "Legacy">;
```

Which is used in CGObjCGNU::GenerateMessageSend

**Comment by davidchisnall** [comment]
> I assume we can split the test into two parts:
> 
> 1. Checking if the driver correctly sets `-fobjc-dispatch-method`
> 2. Checking if frontend emits `objc_msgSend`

Ah, I forgot that was how it propagated. I think we only need the driver tests, there should be existing checks for the dispatch-method thing.

**Comment by hmelder** [comment]
Should be ready now!

**Comment by davidchisnall** [comment]
A lot of formatting changes seem to have crept into unrelated bits of the code, please can you revert them? You can use the clang-format-diff script to format only the bits that you’ve changed.

**Comment by hmelder** [comment]
> A lot of formatting changes seem to have crept into unrelated bits of the code, please can you revert them? You can use > the clang-format-diff script to format only the bits that you’ve changed.

Done. Sorry for the messy commit history.

**Comment by davidchisnall** [comment]
LGTM. @rjmccall, do you want to take a look?

**Review by rjmccall — APPROVED** [review]
LGTM if @davidchisnall's okay with it.

**Review by davidchisnall — APPROVED** [review]

**Pull request merged**